src/Serilog/ReusableStringWriter.cs

```diff
@@ -0,0 +1,47 @@
+﻿namespace Serilog;
+
+/// <summary>
+/// Class that provides reusable StringWriters to reduce memory allocations
+/// </summary>
+public class ReusableStringWriter: StringWriter
+{
+    static readonly object _poolLock = new();
```

Comment:
"Pooling" only a single instance in a `[ThreadStatic]` field might get the same speedup in most cases while avoiding synchronization and the possibility of leaks/annoying pool growth.

Something like a last-in-wins replacement of the cached instance, and then a format provider equality check on retrieval, might deal with the varied-format-provider case reasonably well.

Reply:
1. https://www.nuget.org/packages/Microsoft.Extensions.ObjectPool/
or
2. non-blcoking last-in-wins in-place implementation like [this one](https://github.com/graphql-dotnet/graphql-dotnet/blob/831efff7380f417132f8cdb95accaca4a08994ab/src/GraphQL/Validation/DocumentValidator.cs#L58)

Reply:
Locks+dictionaries+stacks will kill performance.

Reply:
I went with last-in-wins approach with ThreadStatic.